core/unit_test/TestArrayOps.hpp

```diff
@@ -0,0 +1,527 @@
+
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic pop
+#endif
+
+#if defined(__clang__)
+#pragma clang diagnostic pop
+#endif
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+#include <numeric>
+
+namespace {
+
+TEST(TEST_CATEGORY, array_capacity) {
+  using A = Kokkos::Array<int, 2>;
+  A a{{3, 5}};
+
+  ASSERT_FALSE(a.empty());
+  ASSERT_EQ(a.size(), 2u);
+  ASSERT_EQ(a.max_size(), 2u);
+}
+
+enum Enum { EZero, EOne };
+enum EnumBool : bool { EBFalse, EBTrue };
+
+TEST(TEST_CATEGORY, array_element_access) {
+  using A = Kokkos::Array<int, 2>;
+  A a{{3, 5}};
+  A const& ca = a;
+
+  size_t index = 1;
+  ASSERT_EQ(a[index], 5);
+
+  auto b = static_cast<bool>(index);
+  ASSERT_EQ(a[b], a[index]);
+  ASSERT_EQ(ca[b], a[index]);
+
+  auto c = static_cast<char>(index);
+  ASSERT_EQ(a[c], a[index]);
+  ASSERT_EQ(ca[c], a[index]);
+
+  auto sc = static_cast<signed char>(index);
+  ASSERT_EQ(a[sc], a[index]);
+  ASSERT_EQ(ca[sc], a[index]);
+
+  auto uc = static_cast<unsigned char>(index);
+  ASSERT_EQ(a[uc], a[index]);
+  ASSERT_EQ(ca[uc], a[index]);
+
+#if defined(__cpp_char8_t)
+  auto c8 = static_cast<char8_t>(index);
+  ASSERT_EQ(a[ch], a[index]);
+  ASSERT_EQ(a[ch], a[index]);
+#endif
+
+  auto c16 = static_cast<char16_t>(index);
+  ASSERT_EQ(a[c16], a[index]);
+  ASSERT_EQ(ca[c16], a[index]);
+
+  auto c32 = static_cast<char32_t>(index);
+  ASSERT_EQ(a[c32], a[index]);
+  ASSERT_EQ(ca[c32], a[index]);
+
+  auto wc = static_cast<wchar_t>(index);
+  ASSERT_EQ(a[wc], a[index]);
+  ASSERT_EQ(ca[wc], a[index]);
+
+  auto s = static_cast<short>(index);
+  ASSERT_EQ(a[s], a[index]);
+  ASSERT_EQ(ca[s], a[index]);
+
+  auto us = static_cast<unsigned short>(index);
+  ASSERT_EQ(a[us], a[index]);
+  ASSERT_EQ(ca[us], a[index]);
+
+  auto i = static_cast<int>(index);
+  ASSERT_EQ(a[i], a[index]);
+  ASSERT_EQ(ca[i], a[index]);
+
+  auto ui = static_cast<unsigned int>(index);
+  ASSERT_EQ(a[ui], a[index]);
+  ASSERT_EQ(ca[ui], a[index]);
+
+  auto l = static_cast<long>(index);
+  ASSERT_EQ(a[l], a[index]);
+  ASSERT_EQ(ca[l], a[index]);
+
+  auto ul = static_cast<unsigned long>(index);
+  ASSERT_EQ(a[ul], a[index]);
+  ASSERT_EQ(ca[ul], a[index]);
+
+  auto ll = static_cast<long long>(index);
+  ASSERT_EQ(a[ll], a[index]);
+  ASSERT_EQ(ca[ll], a[index]);
+
+  auto ull = static_cast<unsigned long long>(index);
+  ASSERT_EQ(a[ull], a[index]);
+  ASSERT_EQ(ca[ull], a[index]);
+
+  auto e = static_cast<Enum>(index);
+  ASSERT_EQ(a[e], a[index]);
+  ASSERT_EQ(ca[e], a[index]);
+
+  auto eb = static_cast<EnumBool>(index);
+  ASSERT_EQ(a[eb], a[index]);
+  ASSERT_EQ(ca[eb], a[index]);
+
+#if defined(__clang__)
+  if constexpr (std::is_integral_v<__int128>) {
+    auto i128 = static_cast<__int128>(index);
+    ASSERT_EQ(a[i128], a[index]);
+    ASSERT_EQ(ca[i128], a[index]);
+
+    auto u128 = static_cast<unsigned __int128>(index);
+    ASSERT_EQ(a[u128], a[index]);
+    ASSERT_EQ(ca[u128], a[index]);
+  }
+#endif
```

Comment:
We don't use or test `__int128` anywhere else in `Kokkos`. Why do you care about it here?

Reply:
I was trying to have a complete set of `is_integral` tests.

Reply:
I would just drop it. `__int128` isn't portable anyway.

Reply:
Removed.